wrapanapi/base/system.py

```diff
@@ -0,0 +1,166 @@
+    """Represents any system that wrapanapi interacts with."""
+    @property
+    def logger(self):
+        return logging.getLogger(self.__class__.__module__ + "." + self.__class__.__name__)
+
+    def stats(self, *requested_stats):
+        """Returns all available stats, if none are explicitly requested
+
+        Args:
+            *requested_stats: A list giving the name of the stats to return. Stats are defined
+                in the _stats_available attibute of the specific class.
+        Returns: A dict of stats.
+        """
+        if not hasattr(self, '_stats_available'):
+            raise Exception('{} is missing self._stats_available dictionary'.format(
+                self.__class__.__name__))
+
+        requested_stats = requested_stats or self._stats_available
+        return {stat: self._stats_available[stat](self) for stat in requested_stats}
+
+    def disconnect(self):
+        pass
+
+
+class BaseVMSystem(BaseSystem):
+    """
+    Base interface class for Management Systems that manage VMs/instances
+
+    Interface notes:
+
+    * Initializers of subclasses must support \*\*kwargs in their
+      signtures
+
+    """
+    __metaclass__ = ABCMeta
+    STEADY_WAIT_MINS = 3
+
+    @classmethod
+    @abstractproperty
+    def can_suspend(cls):
+        """Indicates whether this system can suspend VM's/instances."""
+        raise NotImplementedError
+
+    @classmethod
+    @abstractproperty
+    def can_pause(cls):
+        """Indicates whether this system can pause VM's/instances."""
+        raise NotImplementedError
+
+    @classmethod
+    @abstractproperty
+    def steady_wait_time(cls):
+        """Returns default secs to wait for VM to move into a steady state."""
+        return 120
+
+    @abstractmethod
+    def list_vms(self, **kwargs):
+        """Returns a list of VM objects
+
+        Returns: list of BaseVM
+        """
+        raise NotImplementedError('list_vm not implemented.')
+
+    @abstractmethod
+    def create_vm(self, vm_name, *args, **kwargs):
+        """Creates a vm.
+
+        Args:
+            vm_name: name of the vm to be created
+        Returns: BaseVM if creation successful
+        """
+        raise NotImplementedError('create_vm not implemented.')
+
+    @abstractmethod
+    def list_templates(self):
+        """Returns a list of templates/images.
+
+        Returns: list of template/image names
+        """
+        raise NotImplementedError('list_template not implemented.')
+
+    @abstractmethod
+    def list_flavors(self):
+        """Returns a list of flavors.
+
+        Only valid for OpenStack and Amazon
+
+        Returns: list of flavor names
+        """
+        raise NotImplementedError('list_flavor not implemented.')
+
+    def list_networks(self):
+        """Returns a list of networks.
+
+        Only valid for OpenStack
+
+        Returns: list of network names
+        """
+        raise NotImplementedError('list_network not implemented.')
+
+    @abstractmethod
```

Comment:
combining abstract methods and not implemented errors is not helpful since abstract methods prevent instantiation if not overridden

perhaps support concepts should be put into mixins

Reply:
That is true. This is the way it had already been written so I stuck to the convention. I guess the only advantage here is if someone super()'s when they should not have. Since it is possible to define an abtract property/method with only a docstring, would we rather just have no code? 

Re mixins, I see what you are thinking. But the System class should mainly have "get_" and "list_" methods that return implementations (or a list of implementations) of BaseEntity. (In fact, I may move 'create' underneath the Entity as well). So do you think splitting System up further into a bunch of mixins that only have a couple methods might be be "over-classifying" things?

I'm thinking it might be simpler to just keep the "optional" get/list methods as non-abstract and raise NotImplementedError on them. For ex "list_flavors" should probably not be an abstractmethod since only 2 providers support that.